packages/nextjs/src/config/webpack.ts

```diff
@@ -86,59 +86,128 @@ export function constructWebpackConfigFunction(
-          // We generally want to apply the loader to all API routes, pages and to the middleware file.
-
-          // `resourcePath` may be an absolute path or a path relative to the context of the webpack config
-          let absoluteResourcePath: string;
-          if (path.isAbsolute(resourcePath)) {
-            absoluteResourcePath = resourcePath;
-          } else {
-            absoluteResourcePath = path.join(projectDir, resourcePath);
-          }
-          const normalizedAbsoluteResourcePath = path.normalize(absoluteResourcePath);
-
-          if (
-            // Match everything inside pages/ with the appropriate file extension
+          const normalizedAbsoluteResourcePath = normalizeLoaderResourcePath(resourcePath);
+          return (
             normalizedAbsoluteResourcePath.startsWith(pagesDirPath) &&
+            !normalizedAbsoluteResourcePath.startsWith(apiRoutesPath) &&
             dotPrefixedPageExtensions.some(ext => normalizedAbsoluteResourcePath.endsWith(ext))
-          ) {
-            return true;
-          } else if (
-            // Match middleware.js and middleware.ts
-            normalizedAbsoluteResourcePath === middlewareJsPath ||
-            normalizedAbsoluteResourcePath === middlewareTsPath
-          ) {
-            return userSentryOptions.autoInstrumentMiddleware ?? true;
-          } else {
-            return false;
-          }
+          );
+        },
+        use: [
+          {
+            loader: path.resolve(__dirname, 'loaders', 'wrappingLoader.js'),
+            options: {
+              ...staticWrappingLoaderOptions,
+              wrappingTargetKind: 'page',
+            },
+          },
+        ],
+      });
+
+      // Wrap api routes
+      newConfig.module.rules.unshift({
+        test: resourcePath => {
+          const normalizedAbsoluteResourcePath = normalizeLoaderResourcePath(resourcePath);
+          return (
+            normalizedAbsoluteResourcePath.startsWith(apiRoutesPath) &&
+            dotPrefixedPageExtensions.some(ext => normalizedAbsoluteResourcePath.endsWith(ext))
+          );
+        },
+        use: [
+          {
+            loader: path.resolve(__dirname, 'loaders', 'wrappingLoader.js'),
+            options: {
+              ...staticWrappingLoaderOptions,
+              wrappingTargetKind: 'api-route',
+            },
+          },
+        ],
+      });
+
+      // Wrap middleware
+      newConfig.module.rules.unshift({
+        test: resourcePath => {
+          const normalizedAbsoluteResourcePath = normalizeLoaderResourcePath(resourcePath);
+          return (
+            normalizedAbsoluteResourcePath === middlewareJsPath || normalizedAbsoluteResourcePath === middlewareTsPath
+          );
+        },
+        use: [
+          {
+            loader: path.resolve(__dirname, 'loaders', 'wrappingLoader.js'),
+            options: {
+              ...staticWrappingLoaderOptions,
+              wrappingTargetKind: 'middleware',
```

Comment:
m: Ideally I would like to see some unit tests that show we are applying the correct `wrappingTargetKind` - this seems fragile and easy to regress on if we change the path test.

Reply:
Added some in 9401d66053a1f5ebafad2fbaf572bab33097a86f

Thanks for holding me accountable to tests btw.